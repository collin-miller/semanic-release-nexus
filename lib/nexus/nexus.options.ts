export interface INexusOptions {
    username?: string;
    password?: string;
    retryStrategy?: (times: number) => number | void | null;
    scheme?: string;
}

export const DEFAULT_NEXUS_OPTIONS: INexusOptions = {
    scheme: 'http://',
    retryStrategy: (times) => {
        return Math.min(times * 50, 2000);
    },
};
