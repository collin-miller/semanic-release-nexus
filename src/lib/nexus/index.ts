import Nexus from './nexus';

export * from './nexus.options';
export default Nexus;
