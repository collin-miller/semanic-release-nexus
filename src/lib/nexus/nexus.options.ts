export interface INexusOptions {
    username?: string;
    password?: string;
    retryStrategy?: (times: number) => number | void | null;
}

export const DEFAULT_NEXUS_OPTIONS: INexusOptions = {
    retryStrategy: (times) => {
        return Math.min(times * 50, 2000);
    },
};
