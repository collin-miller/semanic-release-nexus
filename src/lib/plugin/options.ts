import { IPluginConfig } from './config';

export interface IPluginOptions extends IPluginConfig {
    auth: {
        username?: string;
        password?: string;
    };
}
