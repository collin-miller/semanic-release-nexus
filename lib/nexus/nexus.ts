import axios, { AxiosRequestConfig } from 'axios';
import Debug from 'debug';
import FormData from 'form-data';
import fs from 'fs';
import { INexusOptions, DEFAULT_NEXUS_OPTIONS } from './nexus.options';

const debug = Debug('nexus');

export default class Nexus {
    /**
     * options getter
     */
    public get options(): INexusOptions {
        return this.nexusOptions;
    }

    /**
     * options setter
     */
    public set options(options: INexusOptions) {
        this.nexusOptions = Object.assign({} as INexusOptions, DEFAULT_NEXUS_OPTIONS, options);
    }

    /**
     * constructor
     */
    constructor(public host: string = 'localhost', private nexusOptions: INexusOptions = {} as INexusOptions) {
        this.options = nexusOptions;
    }

    /**
     * deploy
     */
    public async deploy(repo: string, artifactName: string, artifactPath: string): Promise<void> {
        let options = {} as AxiosRequestConfig;

        const form = new FormData();
        form.append(artifactName, fs.createReadStream(artifactPath));
        options = {
            data: form,
            headers: form.getHeaders(),
        };
        if (this.options.username && this.options.password) {
            options.auth = { username: this.options.username, password: this.options.password };
        }
        try {
            await axios.post(`${this.options.scheme}${this.host}/repository/${repo}`, options);
        } catch (error) {
            debug('%o', error);
            throw error;
        }
    }
}
