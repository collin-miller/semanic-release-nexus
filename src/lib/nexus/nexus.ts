import axios, { AxiosRequestConfig } from 'axios';
import Debug from 'debug';
import FormData from 'form-data';
import fs from 'fs';
import { INexusOptions, DEFAULT_NEXUS_OPTIONS } from './nexus.options';

const debug = Debug('nexus');

export default class Nexus {
    /**
     * options getter
     */
    public get options(): INexusOptions {
        return this.nexusOptions;
    }

    /**
     * options setter
     */
    public set options(options: INexusOptions) {
        this.nexusOptions = Object.assign({} as INexusOptions, DEFAULT_NEXUS_OPTIONS, options);
    }

    /**
     * constructor
     */
    constructor(public host: string = 'localhost', private nexusOptions: INexusOptions = {} as INexusOptions) {
        this.options = nexusOptions;
    }

    /**
     * deploy
     */
    public async deploy(pathPrefix: string, artifactName: string, artifactPath: string): Promise<void> {
        const form = new FormData();

        form.append('raw.directory', pathPrefix);
        form.append('raw.asset1', fs.createReadStream(artifactPath));
        form.append('raw.asset1.filename', `${artifactName}`);

        let config = {
            headers: form.getHeaders(),
        } as AxiosRequestConfig;

        config = {
            data: form,
            headers: form.getHeaders(),
        };
        if (this.options.username && this.options.password) {
            config.auth = { username: this.options.username, password: this.options.password };
        }
        try {
            await axios.post(`${this.host}/service/rest/v1/components?repository=raw`, form, config);
        } catch (error) {
            debug('%o', error);
            throw error;
        }
    }
}
