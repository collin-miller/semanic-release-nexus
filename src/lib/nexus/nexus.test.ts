import axios from 'axios';
import { afterEach, beforeEach, describe, expect, jest, it } from '@jest/globals';
import fs from 'fs';
import Nexus from './nexus';
import { INexusOptions } from './nexus.options';

jest.mock('axios');

describe('Nexus Client', () => {
    it('should instantiate with default values', () => {
        const nx = new Nexus();
        expect(nx.host).toEqual('localhost');
        expect(nx.options).toMatchSnapshot();
    });

    it('should work with override values', () => {
        const nx = new Nexus('some-host', {
            apiKey: 'some-api-key',
            retryStrategy: (number) => {
                return number;
            },
        } as INexusOptions);
        expect(nx.host).toEqual('some-host');
        expect(nx.options.retryStrategy?.(1)).toEqual(1);
        expect(nx.options).toMatchSnapshot();
    });

    it('should return correct retry strategy', () => {
        const nx = new Nexus();
        expect(nx.options.retryStrategy?.(2)).toEqual(100);
    });
    describe('upload to nexus', () => {
        const mockedAxios = axios as jest.Mocked<typeof axios>;
        const originalMath = global.Math;
        beforeEach(() => {
            mockedAxios.post.mockReturnValue(Promise.resolve({ success: true }));

            const mockMath = Object.create(global.Math);
            mockMath.random = () => 0.5;
            global.Math = mockMath;
        });

        afterEach(() => {
            global.Math = originalMath;
            jest.restoreAllMocks();
            jest.clearAllMocks();
        });
        it('should upload', async () => {
            const fsMock = jest.spyOn(fs, 'createReadStream');
            const postMock = jest.spyOn(axios, 'post').mockImplementation(() => Promise.resolve({ data: {} }));
            const nx = new Nexus();
            await nx.deploy('some-repo', 'some-artifact-name', '/some/path/to/artifact.zip');

            expect(fsMock).toHaveBeenCalledWith('/some/path/to/artifact.zip');
            expect(postMock).toHaveBeenCalledWith(
                'localhost/repository/some-repo',
                expect.objectContaining({
                    data: expect.any(Object),
                    headers: {
                        'content-type':
                            'multipart/form-data; boundary=--------------------------555555555555555555555555',
                    },
                }),
            );
        });
        it('should upload w/ auth', async () => {
            const fsMock = jest.spyOn(fs, 'createReadStream');
            const postMock = jest.spyOn(axios, 'post').mockImplementation(() => Promise.resolve({ data: {} }));
            const nx = new Nexus('localhost', { username: 'some-user', password: 'some-password' });

            await nx.deploy('some-repo', 'some-artifact-name', '/some/path/to/artifact.zip');

            expect(fsMock).toHaveBeenCalledWith('/some/path/to/artifact.zip');
            expect(postMock).toHaveBeenCalledTimes(1);

            await expect(postMock).toHaveBeenCalledWith(
                'localhost/repository/some-repo',
                expect.objectContaining({
                    data: expect.any(Object),
                    headers: {
                        'content-type':
                            'multipart/form-data; boundary=--------------------------555555555555555555555555',
                    },
                    auth: { password: 'some-password', username: 'some-user' },
                }),
            );
        });
        it('should catch error and raise', async () => {
            const fsMock = jest.spyOn(fs, 'createReadStream');
            const postMock = jest
                .spyOn(axios, 'post')
                .mockImplementation(() => Promise.reject(new Error('Some strange error')));
            const nx = new Nexus();
            await expect(nx.deploy('some-repo', 'some-artifact-name', '/some/path/to/artifact.zip')).rejects.toThrow();

            expect(fsMock).toHaveBeenCalledWith('/some/path/to/artifact.zip');
            expect(postMock).toHaveBeenCalled();
        });
    });
});
