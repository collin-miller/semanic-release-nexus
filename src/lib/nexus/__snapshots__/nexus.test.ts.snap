// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nexus Client should instantiate with default values 1`] = `
Object {
  "retryStrategy": [Function],
  "scheme": "http://",
}
`;

exports[`Nexus Client should work with override values 1`] = `
Object {
  "apiKey": "some-api-key",
  "retryStrategy": [Function],
  "scheme": "http://",
}
`;
